Solutions to Assignment1

andreeapetcu opened the pull request.

andreeapetcu committed:
Create LinkedListElement.java

Q2 - O(n) complexity (the "runner" technique)

andreeapetcu committed:
Q1 solution - anagrams

flerdacodeu commented on andreeapetcu/assignment1/Anagrams.java:
```diff
@@ -0,0 +1,110 @@
+
+public class Anagrams {
+
+    private final int difference = 'a'- 'A';
```
This is a difference, as it is the difference between two value, but it is not clear what it represents.
Maybe adding a comment would help.

flerdacodeu replied:
This is a constant, so you want to declare it as:

```java
  private static final int DIFFERENCE = 'a' - 'A';
```

The `static` means there is a single copy of it and the UPPER CASE name is a way to highlight to the reader that this is a constant (a convention, but a quite useful one, because the reader can then immediately know the value will not change or depend on the instance).

flerdacodeu commented on andreeapetcu/assignment1/Anagrams.java:
```diff
@@ -0,0 +1,110 @@
+    private final String splitExpression = "\\s*[^a-zA-Z0-9]+\\s*";
+
+
+    private char caseInsensitive(char c){
```
What you are trying to do here is actually turning a character to lower case.
There is already a handy function to convert an entire string to lower case in one go: `String.toLowerCase()`.
If there is a function in the library that does what you need, it is usually best to use it (and it comes with extra benefits, like supporting non-ASCII characters in case you want to extend your implementation at a later time).

I know it is hard to know if there is such a function you can reuse (it is difficult for me at times too, and code reviews help with that too!). It also comes with experience, so as you become a more experienced software engineer you yourself will be able to see more of these cases without being told.

flerdacodeu commented on andreeapetcu/assignment1/Anagrams.java:
```diff
@@ -0,0 +1,110 @@
+     */
+    public boolean wordAnagrams(String input1, String input2, boolean caseSensitive){
+        //contains count of each letter in the first word
+        int[] letters = new int[128];
```
You might want to explain the choice of `128` in a comment.

flerdacodeu replied:
The comment above says this is a count, but the name `letters` does not make that clear. For example, it could be an array containing the letters in the first word. How about `lettersCount1`?

flerdacodeu commented on andreeapetcu/assignment1/Anagrams.java:
```diff
@@ -0,0 +1,110 @@
+        boolean[] checkedWordsSentence1 = new boolean[splitSentence1.length];
+
+        //words from sentence 2 that have a pair in sentence 1
+        boolean[] checkedWordsSentence2 = new boolean[splitSentence2.length];
```
Well done! It is important to make sue that all words in both sentences have an equivalent in the other one.
There is still a possible issue, however. What if one word has two possible anagrams and they both appear in the sentence?

Note that this problem is similar to the one for characters within a word, where you needed counts to make sure that `"aab"` and `"baa"` would not be considered anagrams (all letters in the first have an equivalent in the second one and viceversa, but the counts do not match).

Does it make sense? Can you see how to fix it?

andreeapetcu replied:
Are you talking about cases where sentence1 = "At least he did not steal my book of fairy tales" and sentence2 = "at tales did not he least my book of steal fairy" (the word "least" has two anagrams that appear in the sentence: steal and tales)?

In this case, my algorithm looks for an anagram of "least" and the first match in the second sentence is "tales"; it now sets the corresponding position of "least" in checkWordsSentence1 to true and the corresponding position of "tales" in checkedWordsSentence2 to true. When the iteration in the first sentence gets to "steal", the first anagram in the second sentence is again "tales"; however, since "tales" is already set to true, the iteration continues and the next anagram of "steal" that is set to false in the second sentence, "least" in my example, will be the match. Should the algorithm not find a matching anagram in the second sentence for "steal", then "steal" will remain false and the for loop that checks if every entry of checkedWordsSentence1 and checkedWordsSentence2 is true, will return false.

If this is not the issue, could you please give me another example?

flerdacodeu replied:
You are right! I had not understood correctly the additional check in the `if`-statement.
One possible style change to make this easier to understand:

```java
        for (int i = 0; i< splitSentence1.length; i++) {
            for (int j = 0; j < splitSentence2.length; j++) {
                if (checkedWordsSentence2[j]) {
                    // This word has already been used as anagram of a previous word
                    // in the sentence.
                    continue;
                }
                if (wordAnagrams(splitSentence2[j], splitSentence1[i], caseSensitive)) {
                    checkedWordsSentence1[i] = true;
                    checkedWordsSentence2[j] = true;
                    break;
                }
            }
        }
```

flerdacodeu commented on andreeapetcu/assignment1/LinkedListElement.java:
```diff
@@ -0,0 +1,59 @@
+public class LinkedListElement {
+
+    public class LinkedListNode<T> {
+        LinkedListNode<T> next = null;
```
Should this also be marked as `public`?

flerdacodeu commented on andreeapetcu/assignment1/LinkedListElement.java:
```diff
@@ -0,0 +1,59 @@
+        }
+
+        /**
+         * Adds a node at the end of the linked list (used for testing method)
```
I would not say `(used for testing method)` in this javadoc.
Having a method to append at the end of the list seems like a reasonable thing to have.
Think of building this solution for someone else to use.

flerdacodeu commented on andreeapetcu/assignment1/LinkedListElement.java:
```diff
@@ -0,0 +1,59 @@
+         * @param value the value of the node to be appended
+         */
+        public void addNode (T value){
+            LinkedListNode<T> newNode = new LinkedListNode<>(value);
```
Here you construct a new node that you might actually not need.
It seems unnecessary and it would create some garbage to be collected at a later time, which is wasteful.

flerdacodeu commented on andreeapetcu/assignment1/LinkedListElement.java:
```diff
@@ -0,0 +1,59 @@
+            LinkedListNode<T> newNode = new LinkedListNode<>(value);
+            LinkedListNode<T> node = this;
+
+            if (node == null)
```
Can node actually ever be `null` in this case? What is it initialised with?

andreeapetcu replied:
No, 'node' cannot be null; the keyword "this" initializes it with the head of the list

flerdacodeu commented on andreeapetcu/assignment1/LinkedListElement.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    /**
+     * Method that finds the kth to last element of a linked list
```
Style: avoid saying `Method that` or `Class that`.
Instead for methods, just says:

```java
  /**
   * Finds the k-th to last element of a linked list.
   *
   ...
```

flerdacodeu commented on andreeapetcu/assignment1/LinkedListElement.java:
```diff
@@ -0,0 +1,59 @@
+        LinkedListNode<Integer> normalPointer = head;
+
+        if (head == null)
+            return normalPointer;
```
It would be clearer here to just return `null` directly.

Right now, to understand that this code does I need to notice that `normalPointer` is initialised to `head` and the `if`-statement makes sure that `head` is `null`, so `normalPointer` is `null` in this case.

flerdacodeu commented on andreeapetcu/assignment1/LinkedListElement.java:
```diff
@@ -0,0 +1,59 @@
+        if (head == null)
+            return normalPointer;
+
+        LinkedListNode<Integer> fastPointer = head.next;
```
I would initialise these two references together and maybe add a comment about how they are meant to work together. (you want them to be `k+1` elements apart, so that one reaches the end of the list, the other is `k` elements from the end).

```java
  // Point two elements in the list that are k+1 elements apart, so that when fastPointer goes
  // beyond the end of the list, normalPointer is at the k-th to last element.
```

flerdacodeu reviewed (changes requested):
Thanks for your submission.

I have a few comments for you to address. Feel free to reach out to me on Slack or reply on GitHub if anything is unclear. Remember this is a conversation to make the results better, not grading, so take each comment as an opportunity to improve (I get many comments on my own code all the time).

andreeapetcu committed:
Update Anagrams.java

andreeapetcu committed:
Update LinkedListElement.java

andreeapetcu committed:
Refactored the linked list data structure

Created a separate inner class that defines a node of a liked list (moved the method for adding nodes to the list in the outer class); 
The inner class is enclosed in the class that defines basic functionality of a linked list

andreeapetcu commented:
Thank you for reviewing my code and suggesting improvements! :) I changed my code to reflect your suggestions.

flerdacodeu commented on andreeapetcu/assignment1/Anagrams.java:
```diff
@@ -0,0 +1,103 @@
+
+public class Anagrams {
+
+    private final String splitExpression = "\\s*[^a-zA-Z0-9]+\\s*";
```
Style: This is also a constant, so you can define it as `static` and make it `UPPER_CASE`.

flerdacodeu commented on andreeapetcu/assignment1/Anagrams.java:
```diff
@@ -0,0 +1,103 @@
+     * @param input1 first word
+     * @param input2 second word
+     * @param caseSensitive flag for case sensitive or case insensitive anagrams
+     *
```
Remove this "blank" line.

flerdacodeu commented on andreeapetcu/assignment1/Anagrams.java:
```diff
@@ -0,0 +1,103 @@
+     * @param caseSensitive flag for case sensitive or case insensitive anagrams
+     *
+     */
+    public boolean wordAnagrams(String input1, String input2, boolean caseSensitive){
```
Style: it is common practice to add a space before `{`, like on this line (but same applies in a few other places).

flerdacodeu commented on andreeapetcu/assignment1/Anagrams.java:
```diff
@@ -0,0 +1,103 @@
+        // array size equals the number of ASCII characters
+        int[] lettersCount = new int[128];
+
+        if (input1.length() != input2.length())
```
Style: the Google style guide suggests always putting `{` and `}` around `if`- and `for`-statement code blocks, even if there is a single statement.

This is mostly to avoid accidentally adding a line within what looks like the code block but, because there are no `{` and `}` around it, making it incorrect.

flerdacodeu commented on andreeapetcu/assignment1/Anagrams.java:
```diff
@@ -0,0 +1,103 @@
+        if (input1.length() != input2.length())
+            return false;
+
+        char currentLetter;
```
Style: While you use `currentLetter` within both `for`-loops, it is actually better to define it within each loop body. Generally, reducing the scope of a variable is a good idea. And the compiler will not "waste" space for it, because it knows where it is used.

However, as it is now, it is possible to accidentally use the value of the variable outside the loop, which would not be very meaningful.

flerdacodeu commented on andreeapetcu/assignment1/Anagrams.java:
```diff
@@ -0,0 +1,110 @@
+                    checkedWordsSentence2[j] = true;
+                }
+            }
+        }
```
Optional: if you get at the end of the inner loop and you did not find an anagram for word `i`, then you could terminate right away without having to check all of the rest.

flerdacodeu commented on andreeapetcu/assignment1/Anagrams.java:
```diff
@@ -0,0 +1,110 @@
+        }
+
+        for (int i =0; i< splitSentence1.length; i++)
+            if (!checkedWordsSentence1[i] || !checkedWordsSentence2[i])
```
Optional: Is it possible to have that `checkedWordsSentence1` contains `true` but there is a `false` within `checkedWordsSentence2`?

flerdacodeu commented on andreeapetcu/assignment1/LinkedList.java:
```diff
@@ -0,0 +1,98 @@
+     * @param k position of the element starting from the end of the list
+     * @return kth to last element of the list
+     */
+    public LinkedList<T> kthToLastElement(int k){
```
I think this function should return `T` rather than a `LinkedList<T>`, i.e., return the element itself.

flerdacodeu commented on andreeapetcu/assignment1/LinkedList.java:
```diff
@@ -0,0 +1,98 @@
+    /**
+     * Traverses the list
+     */
+    public void traverse(){
```
The purpose of this method (as I see it), is to print the content of the list.
Maybe call this `print()` rather than `traverse()`.

flerdacodeu reviewed (approved):
Thanks for the updates!
Just a few style suggestions.

andreeapetcu committed:
Update Anagrams.java

andreeapetcu committed:
Update LinkedList.java
